prebuilt-ui/basic-embed/components/PrebuiltCall.js

```diff
@@ -0,0 +1,254 @@
+  const callFrame = useRef(null);
+  const [linkCopied, setLinkCopied] = useState(false);
+
+  // Updates the time left that displays in the UI
+  useEffect(() => {
+    if (!exp) {
+      return false;
+    }
+    const i = setInterval(() => {
+      const timeLeft = Math.floor((new Date(exp * 1000) - Date.now()) / 1000);
+      setSecs(`${Math.floor(timeLeft / 60)}:${`0${timeLeft % 60}`.slice(-2)}`);
+    }, 1000);
+    return () => clearInterval(i);
+  }, [exp]);
+
+  // Listens for a "call" demo state, and creates then joins a callFrame as soon as that happens
+  useEffect(() => {
+    if (!iframeRef?.current || demoState !== 'call') {
+      return;
+    }
+    try {
+      callFrame.current = DailyIframe.createFrame(iframeRef.current, {
+        showLeaveButton: true,
+        iframeStyle: {
+          height: '100%',
+          width: '100%',
+          aspectRatio: 16 / 9,
+          minwidth: '400px',
+          maxWidth: '920px',
+          border: '0',
+          borderRadius: '12px',
+        },
+      });
+      callFrame.current.join({
+        url: roomURL,
+      });
+    } catch (e) {
+      setDemoState('home');
+      setIsError(true);
+      return;
+    }
+
+    const handleLeftMeeting = () => {
+      setDemoState('home');
+      setRoomValidity(false);
+      callFrame.current.destroy();
+    };
+
+    callFrame.current.on('left-meeting', handleLeftMeeting);
+  }, [demoState, iframeRef, roomURL]);
+
+  const createRoom = useCallback(
+    async (e) => {
+      if (!roomURLRef?.current.value) {
+        const roomExp = Math.round(Date.now() / 1000) + 60 * 5;
+        try {
+          const res = await fetch('/api/room', {
+            method: 'POST',
+            headers: {
+              Accept: 'application/json',
+              'Content-Type': 'application/json',
+            },
+            body: JSON.stringify({
+              properties: {
+                roomExp,
+              },
+            }),
+          });
+          const roomJson = await res.json();
+          const { url } = roomJson;
+          setRoomURL(url);
+          setDemoState('call');
+          setExp(roomExp);
+          setIsError(false);
+        } catch (e) {
+          setDemoState('home');
+          setIsError(true);
+        }
+      }
+    },
+    [roomURL, exp]
+  );
+
+  // Updates state if a room is provided
+  const handleRoomInput = useCallback(
+    (e) => {
+      setRoomURL(e?.target?.value);
+      if (e?.target?.checkValidity()) {
+        setRoomValidity(true);
+        console.log(roomValidity);
+      }
+    },
+    [roomValidity]
+  );
+
+  const submitJoinRoom = useCallback(() => {
+    setDemoState('call');
+  });
+
+  const handleCopyClick = useCallback(() => {
+    console.log('click');
+    writeText(roomURL);
+    setLinkCopied(true);
+    setTimeout(() => setLinkCopied(false), 5000);
+  }, [roomURL, linkCopied]);
+
+  const content = useMemo(() => {
+    switch (demoState) {
+      case 'home':
```

Comment:
The reason why I was keeping this all in one place is because I wanted to demonstrate how basic this basic embed could be and minimize state management complexity. Moving it out means refactoring to add in context and potentially providers (I think), which doesn't help the "basic" narrative IMO. But that is how we've done things in other demos, so I can move!

Reply:
I'm going to push back even more than if that's the reasoning then because we can simplify this quite a bit. Two things:
- if the code's turning out to not be super basic, then putting it all in one component will make it feel even more complicated to read than breaking it in smaller components. If there's a bunch of jsx, it's a flag you can usually break it down a bit and pass props for any dynamic info from the parent or move that logic to the sub-component if it's only used in the sub-component (e.g. Home or Call). You don't need contexts or providers to accommodate this change, though, so don't worry about that.
- you don't actually need to memoize this because the `demoState` isn't changing a bunch so you can simplify even further by removing the `useMemo` and switch statement and just have something like this in component's return:
`{demoState === 'home' ?<Home /> : <Call/>}`

Reply:
Agreed, thank you so much for the insights! I saw this (more components > easier) in action reading your blog today. I've taken a first pass at component-izing that I'll soon request your review on (including a separate component for the API Routes request), so I'll resolve this.